core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,318 @@ private RexNode simplifyCoalesce(RexCall call) {
-          newOperands.add(operands.get(++i));
-          if (unknownAs == FALSE && RexUtil.isNull(operands.get(i))) {
-            values.add(rexBuilder.makeLiteral(false).toString());
+    List<CaseBranch> inputBranches =
+        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList(call.getOperands()));
+
+    // run simplification on all operands
+    RexSimplify branchSimplifier = this;
+    RelDataType branchType = call.getType();
+
+    List<CaseBranch> branches = new ArrayList<>();
+    for (CaseBranch branch : inputBranches) {
+      // simplify the condition
+      RexNode newCond = branchSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+
+      // use the condition to simplify the branch
+      RexNode value = branch.value;
+      RexNode newValue = branchSimplifier.simplify(value, unknownAs);
+
+      branches.add(new CaseBranch(newCond, newValue));
+    }
+
+    // remove branches with invalid conditions
+    branches.removeIf(branch -> branch.cond.isAlwaysFalse());
+
+    // delete all branches after the first AlwaysTrue
+    for (int i = 0; i < branches.size(); i++) {
+      CaseBranch branch = branches.get(i);
+      if (branch.cond.isAlwaysTrue()) {
+        branches.subList(i + 1, branches.size()).clear();
+        break;
+      }
+    }
+
+    branches = compactBranchesWithTheSameConclusion(branches);
+
+    // collect cardinality of values
+    Set<String> values =
+        branches.stream().map(branch -> branch.value.toString()).collect(Collectors.toSet());
+
+    if (values.size() == 1) {
+      final RexNode firstValue = branches.get(0).value;
+
+      if (sameTypeOrNarrowsNullability(branchType, firstValue.getType())) {
+        return firstValue;
+      } else {
+        return rexBuilder.makeAbstractCast(branchType, firstValue);
+      }
+    }
+
+    if (call.getType().getSqlTypeName() == SqlTypeName.BOOLEAN) {
+      final RexNode result = simplifyBooleanCase(rexBuilder, branches, unknownAs, branchType);
+      if (result != null) {
+        if (sameTypeOrNarrowsNullability(branchType, result.getType())) {
+          return simplify(result, unknownAs);
+        } else {
+          // If the simplification would widen the nullability
+          RexNode simplified = simplify(result, UNKNOWN);
+          if (!simplified.getType().isNullable()) {
+            return simplified;
           } else {
-            values.add(operands.get(i).toString());
+            return rexBuilder.makeCast(call.getType(), simplified);
           }
-          break;
-        } else if (operand.isAlwaysFalse() || RexUtil.isNull(operand)) {
-          // Predicate is always FALSE or NULL. Skip predicate and value.
-          ++i;
-          continue;
         }
+      }
+    }
+    List<RexNode> newOperands = CaseBranch.toCaseOperands(rexBuilder, branches);
+    if (newOperands.equals(call.getOperands())) {
+      return call;
+    }
+    return call.clone(call.getType(), newOperands);
```

Comment:
Resulting type can be tightened.
For instance
`case when false then null when x then true else false end` is nullable, however it could be simplified to `case when x then true else false end` which is non-nullable.

Reply:
Great idea!